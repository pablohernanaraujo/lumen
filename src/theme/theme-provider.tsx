import React, {
  createContext,
  type FC,
  useCallback,
  useContext,
  useMemo,
  useState,
} from 'react';
import { useColorScheme } from 'react-native';

import { colorTokens } from './colors';
import { baseTokens } from './tokens';
import type {
  Mode,
  Theme,
  ThemeContextValue,
  ThemeProviderProps,
} from './types';

const ThemeContext = createContext<ThemeContextValue | null>(null);

export const ThemeProvider: FC<ThemeProviderProps> = ({
  children,
  preference,
  persistenceKey = '@lumen-theme',
}) => {
  const systemMode = useColorScheme() as Mode | null;
  const [userMode, setUserMode] = useState<Mode | null>(preference || null);

  const currentMode: Mode = userMode ?? systemMode ?? 'light';

  const theme = useMemo<Theme>(
    () => ({
      ...baseTokens,
      colors: colorTokens[currentMode],
      mode: currentMode,
    }),
    [currentMode],
  );

  const setMode = useCallback((mode: Mode): void => {
    setUserMode(mode);
  }, []);

  const toggleMode = useCallback((): void => {
    setUserMode(currentMode === 'light' ? 'dark' : 'light');
  }, [currentMode]);

  const contextValue = useMemo<ThemeContextValue>(
    () => ({
      theme,
      mode: currentMode,
      setMode,
      toggleMode,
    }),
    [theme, currentMode, setMode, toggleMode],
  );

  return (
    <ThemeContext.Provider value={contextValue}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useTheme = (): ThemeContextValue => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
};
