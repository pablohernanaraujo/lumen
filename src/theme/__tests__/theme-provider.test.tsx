/* eslint-disable max-nested-callbacks */
import React from 'react';
import { Text, useColorScheme } from 'react-native';
import { fireEvent, render } from '@testing-library/react-native';

import { ThemeProvider, useTheme } from '../theme-provider';

// Mock useColorScheme
jest.mock('react-native/Libraries/Utilities/useColorScheme');

const mockUseColorScheme = useColorScheme as jest.MockedFunction<
  typeof useColorScheme
>;

const TestComponent: React.FC = () => {
  const { theme, mode, setMode, toggleMode } = useTheme();

  return (
    <>
      <Text testID="current-mode">{mode}</Text>
      <Text testID="background-color">{theme.colors.background}</Text>
      <Text testID="text-color">{theme.colors.text.primary}</Text>
      <Text testID="set-light" onPress={() => setMode('light')}>
        Set Light
      </Text>
      <Text testID="set-dark" onPress={() => setMode('dark')}>
        Set Dark
      </Text>
      <Text testID="toggle-mode" onPress={toggleMode}>
        Toggle Mode
      </Text>
    </>
  );
};

describe('ThemeProvider', () => {
  beforeEach(() => {
    mockUseColorScheme.mockClear();
  });

  describe('default behavior', () => {
    it('should use system theme when no preference is provided', () => {
      mockUseColorScheme.mockReturnValue('dark');

      const { getByTestId } = render(
        <ThemeProvider>
          <TestComponent />
        </ThemeProvider>,
      );

      expect(getByTestId('current-mode').children[0]).toBe('dark');
    });

    it('should default to light mode when system theme is null', () => {
      mockUseColorScheme.mockReturnValue(null);

      const { getByTestId } = render(
        <ThemeProvider>
          <TestComponent />
        </ThemeProvider>,
      );

      expect(getByTestId('current-mode').children[0]).toBe('light');
    });
  });

  describe('with preference prop', () => {
    it('should use provided preference over system theme', () => {
      mockUseColorScheme.mockReturnValue('dark');

      const { getByTestId } = render(
        <ThemeProvider preference="light">
          <TestComponent />
        </ThemeProvider>,
      );

      expect(getByTestId('current-mode').children[0]).toBe('light');
    });

    it('should render correctly with dark preference', () => {
      const { toJSON } = render(
        <ThemeProvider preference="dark">
          <TestComponent />
        </ThemeProvider>,
      );

      expect(toJSON()).toMatchSnapshot();
    });
  });

  describe('mode manipulation', () => {
    it('should allow setting mode to light', () => {
      const { getByTestId } = render(
        <ThemeProvider preference="dark">
          <TestComponent />
        </ThemeProvider>,
      );

      expect(getByTestId('current-mode').children[0]).toBe('dark');

      fireEvent.press(getByTestId('set-light'));

      expect(getByTestId('current-mode').children[0]).toBe('light');
    });

    it('should allow setting mode to dark', () => {
      const { getByTestId } = render(
        <ThemeProvider preference="light">
          <TestComponent />
        </ThemeProvider>,
      );

      expect(getByTestId('current-mode').children[0]).toBe('light');

      fireEvent.press(getByTestId('set-dark'));

      expect(getByTestId('current-mode').children[0]).toBe('dark');
    });

    it('should toggle mode correctly from light to dark', () => {
      const { getByTestId } = render(
        <ThemeProvider preference="light">
          <TestComponent />
        </ThemeProvider>,
      );

      expect(getByTestId('current-mode').children[0]).toBe('light');

      fireEvent.press(getByTestId('toggle-mode'));

      expect(getByTestId('current-mode').children[0]).toBe('dark');
    });

    it('should toggle mode correctly from dark to light', () => {
      const { getByTestId } = render(
        <ThemeProvider preference="dark">
          <TestComponent />
        </ThemeProvider>,
      );

      expect(getByTestId('current-mode').children[0]).toBe('dark');

      fireEvent.press(getByTestId('toggle-mode'));

      expect(getByTestId('current-mode').children[0]).toBe('light');
    });
  });

  describe('theme colors', () => {
    it('should provide correct light theme colors', () => {
      const { getByTestId } = render(
        <ThemeProvider preference="light">
          <TestComponent />
        </ThemeProvider>,
      );

      const backgroundElement = getByTestId('background-color');
      const textElement = getByTestId('text-color');

      expect(backgroundElement.children[0]).toBeTruthy();
      expect(textElement.children[0]).toBeTruthy();
    });

    it('should provide correct dark theme colors', () => {
      const { getByTestId } = render(
        <ThemeProvider preference="dark">
          <TestComponent />
        </ThemeProvider>,
      );

      const backgroundElement = getByTestId('background-color');
      const textElement = getByTestId('text-color');

      expect(backgroundElement.children[0]).toBeTruthy();
      expect(textElement.children[0]).toBeTruthy();
    });
  });

  describe('error handling', () => {
    it('should throw error when useTheme is used outside ThemeProvider', () => {
      // Capture console.error to prevent cluttering test output
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      expect(() => {
        render(<TestComponent />);
      }).toThrow('useTheme must be used within a ThemeProvider');

      consoleSpy.mockRestore();
    });
  });

  describe('custom persistence key', () => {
    it('should render correctly with custom persistence key', () => {
      const { toJSON } = render(
        <ThemeProvider persistenceKey="@custom-theme">
          <TestComponent />
        </ThemeProvider>,
      );

      expect(toJSON()).toMatchSnapshot();
    });
  });
});
