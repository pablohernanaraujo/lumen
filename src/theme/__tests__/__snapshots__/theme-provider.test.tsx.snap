// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider custom persistence key should render correctly with custom persistence key 1`] = `
[
  <Text
    testID="current-mode"
  >
    dark
  </Text>,
  <Text
    testID="background-color"
  >
    #000000
  </Text>,
  <Text
    testID="text-color"
  >
    #FFFFFF
  </Text>,
  <Text
    onPress={[Function]}
    testID="set-light"
  >
    Set Light
  </Text>,
  <Text
    onPress={[Function]}
    testID="set-dark"
  >
    Set Dark
  </Text>,
  <Text
    onPress={[Function]}
    testID="toggle-mode"
  >
    Toggle Mode
  </Text>,
]
`;

exports[`ThemeProvider with preference prop should render correctly with dark preference 1`] = `
[
  <Text
    testID="current-mode"
  >
    dark
  </Text>,
  <Text
    testID="background-color"
  >
    #000000
  </Text>,
  <Text
    testID="text-color"
  >
    #FFFFFF
  </Text>,
  <Text
    onPress={[Function]}
    testID="set-light"
  >
    Set Light
  </Text>,
  <Text
    onPress={[Function]}
    testID="set-dark"
  >
    Set Dark
  </Text>,
  <Text
    onPress={[Function]}
    testID="toggle-mode"
  >
    Toggle Mode
  </Text>,
]
`;
